test/test_util_network_ssl.cpp

```diff
@@ -676,6 +677,99 @@ TEST(Util_Network_SSL_BasicSendAndReceive)
 }
 
 
+#if REALM_HAVE_SECURE_TRANSPORT
+
+static void run_ssl_nonzero_length_test(
+    test_util::unit_test::TestContext& test_context,
+    std::function<void(test_util::unit_test::TestContext&, std::error_code, std::size_t)>&& read_handler,
+    std::function<void(test_util::unit_test::TestContext&, network::ssl::Stream&)>&& read_error_callback)
+{
+    network::Service service;
+    network::DeadlineTimer run_timer{service};
+    network::Socket socket_1{service};
+    network::Socket socket_2{service};
+    network::ssl::Context ssl_context_1;
+    network::ssl::Context ssl_context_2;
+    configure_server_ssl_context_for_test(ssl_context_1);
+    network::ssl::Stream ssl_stream_1 = {socket_1, ssl_context_1, network::ssl::Stream::server};
+    network::ssl::Stream ssl_stream_2 = {socket_2, ssl_context_2, network::ssl::Stream::client};
+    ssl_stream_1.set_logger(test_context.logger.get());
+    ssl_stream_2.set_logger(test_context.logger.get());
+    connect_ssl_streams(ssl_stream_1, ssl_stream_2);
+    network::ReadAheadBuffer rab;
+
+    char buffer[50];
+
+    auto service_thread = std::thread([&test_context, &service, &run_timer]() {
+        run_timer.async_wait(std::chrono::seconds(10), [&test_context](Status status) {
+            if (!status.is_ok())
+                return;
+            test_context.logger->info("run_ssl_nonzero_length_test: service timed out");
+            abort(); // fail the test if the timer expires
+        });
+        service.run();
+    });
+
+    auto [r_promise, read_future] = util::make_promise_future<void>();
+    auto async_read_handler = [&test_context, read_promise = std::move(r_promise),
+                               handler = std::move(read_handler)](std::error_code ec, std::size_t n) mutable {
+        CHECK_NOT_EQUAL(n, 50);
+        handler(test_context, ec, n);
+        read_promise.emplace_value();
+    };
+
+    // Set the error before the read
+    read_error_callback(test_context, ssl_stream_2);
+    ssl_stream_2.async_read(buffer, 50, rab, std::move(async_read_handler));
```

Comment:
If you did
```
service.post([&] {
    read_error_callback(...);
    ssl_stream_2.async_read(buffer, 50, rab, std::move(async_read_handler));
});
read_future.get();
```

Would you need to interrupt the service in the case where the I/O completion is happening "early"?